### PR TITLE
Refactor(eos_designs): Underlay/Overlay internal logic refactor to better support future MPLS interworking abstraction.

### PR DESCRIPTION
**Pull request by emilarista**
## Change Summary

Refactoring overlay logic to better support future MPLS interworking abstraction.

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes
Added switch.overlay dict to eos_designs_facts.py + removed some unneeded stuff that was replaced by this overlay dict. Modified overlay templates to use the new dict, modified eos_designs_unit_tests vars for the bgp peer group struct_config test to retain test functionality.

## How to test
Tested with molecule, no unintended changes.

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_designs/defaults/main/default-node-type-keys.yml [hunk 1]
```diff
@@ -50,6 +51,7 @@ default_node_type_keys:
       type: rr
       mpls_lsr: true
       default_mpls_overlay_role: server
+      default_evpn_role: server
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/defaults/main/default-node-type-keys.yml:55]: should we set the fabric encap to mpls in this design as well?

**Review by ClausHolbechArista — COMMENTED** [review]
Maybe this should be a feat PR since we add the knob to configure fabric_evpn_encapsulation ?

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py [hunk 2]
```diff
@@ -26,10 +26,8 @@ def foo(self):
         "switch.foo" fact set based on "bar" data model
         """
         return get(self._hostvars, "bar.foo", default="zoo")
-
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]: please revert the unrelated whitespace changes

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py [hunk 3]
```diff
@@ -1624,7 +1604,90 @@ def overlay_routing_protocol_address_family(self):
...
+            and self.underlay_router
+            and self.uplink_type == "p2p"
+        )
+
+        return {"bgp": bgp, "mpls": mpls, "ospf": ospf, "isis": isis}
+
+    @cached_property
+    def overlay(self):
+        """
+        Set overlay.peering_address
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]: Please use a single docstring at the top explaining the output data and meaning, and then use one-liner # above each section.

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py [hunk 4]
```diff
@@ -1624,7 +1604,90 @@ def overlay_routing_protocol_address_family(self):
         return overlay_routing_protocol_address_family
 
     @cached_property
-    def overlay_routing_protocol_peering_address(self):
+    def bgp(self):
+        return (
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py:1623]: please add a docstring to explain the meaning/purpose of the new facts.
```suggestion
        """
        Boolean telling if BGP Routing should be configured
        """
        return (
```

Comment (emilarista) [reply ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py:1623]: Fixed.

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py (hunk 2 above)
Comment (emilarista) [reply ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]: Fixed.

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py (hunk 3 above)
Comment (emilarista) [reply ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]: Fixed.

File: ansible_collections/arista/avd/roles/eos_designs/defaults/main/default-node-type-keys.yml (hunk 1 above)
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/defaults/main/default-node-type-keys.yml:55]: Fixed.

**Review by ClausHolbechArista — APPROVED** [review]

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py [hunk 5]
```diff
@@ -1624,7 +1616,100 @@ def overlay_routing_protocol_address_family(self):
...
+                "isis",
+                "isis-sr",
+                "isis-ldp",
+                "isis-sr-ldp",
+            ]
+            and self.underlay_router
+            and self.uplink_type == "p2p"
+        )
+
+        return {"bgp": bgp, "mpls": mpls, "ospf": ospf, "isis": isis}
```
Comment (tgodaA) [inline ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]:
```suggestion
        if self.uplink_type != "p2p":
            return {"bgp": False, "mpls": False, "ospf": False, "isis": False}
        bgp = self.bgp and self.underlay_routing_protocol == "ebgp" and self.underlay_router
        mpls = (
            self.underlay_routing_protocol in ["isis-sr", "isis-ldp", "isis-sr-ldp", "ospf-ldp"]
            and self.underlay_router
            and self.mpls_lsr
        )
        ospf = self.underlay_routing_protocol in ["ospf", "ospf-ldp"] and self.underlay_router
        isis = (
            self.underlay_routing_protocol
            in [
                "isis",
                "isis-sr",
                "isis-ldp",
                "isis-sr-ldp",
            ]
            and self.underlay_router
        )

        return {"bgp": bgp, "mpls": mpls, "ospf": ospf, "isis": isis}
```

Comment (tgodaA) [reply ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]: Just added negative case: if self.uplink_type != "p2p"

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py [hunk 6]
```diff
@@ -1624,7 +1616,96 @@ def overlay_routing_protocol_address_family(self):
...
+        Toggle to set EVPN encapsulation based on fabric_evpn_encapsulation and node default_evpn_encapsulation.
+        """
+        return get(self._hostvars, "fabric_evpn_encapsulation", default=get(self._node_type_key_data, "default_evpn_encapsulation", default="vxlan"))
+
+    @cached_property
+    def underlay(self):
+        """
+        Returns a dictionary of underlay parameters to configure on the node.
+        """
+        if self.uplink_type != "p2p":
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]: Include `or not self.underlay_router` so you can remove that from all the other tests.

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py [hunk 7]
```diff
@@ -1624,7 +1616,96 @@ def overlay_routing_protocol_address_family(self):
...
+        bgp = self.bgp and self.underlay_routing_protocol == "ebgp" and self.underlay_router
+        mpls = self.underlay_routing_protocol in ["isis-sr", "isis-ldp", "isis-sr-ldp", "ospf-ldp"] and self.underlay_router and self.mpls_lsr
+        ospf = self.underlay_routing_protocol in ["ospf", "ospf-ldp"] and self.underlay_router
+        isis = (
+            self.underlay_routing_protocol
+            in [
+                "isis",
+                "isis-sr",
+                "isis-ldp",
+                "isis-sr-ldp",
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]: remove the trailing comma here. It should make "black" unfold the line like the above.

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py (hunk 6 above)
Comment (emilarista) [reply ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]: Fixed.

File: ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py (hunk 7 above)
Comment (emilarista) [reply ansible_collections/arista/avd/plugins/plugin_utils/eos_designs_facts.py]: Fixed.

File: ansible_collections/arista/avd/roles/eos_designs/templates/overlay/ibgp/peer-groups.j2 [hunk 8]
```diff
@@ -1,23 +1,37 @@
...
       maximum_routes: 0
-{% if switch.evpn_role == "server" or switch.mpls_overlay_role == "server" %}
+{%     if switch.evpn_role == "server" %}
       route_reflector_client: true
+{%     endif %}
+      struct_cfg: {{ overlay_data.evpn_overlay_peer_group.structured_config | arista.avd.default }}
 {% endif %}
-      struct_cfg: {{ overlay_data.overlay_peer_group.structured_config | arista.avd.default }}
-{% if switch.mpls_overlay_role == "server" and overlay_data.mpls_rr_peers is arista.avd.defined %}
+{% if (switch.mpls_overlay_role == "server" or (switch.evpn_role == "server" and switch.overlay.evpn_mpls)) and overlay_data.mpls_rr_peers is arista.avd.defined %}
```
Comment (tgodaA) [inline ansible_collections/arista/avd/roles/eos_designs/templates/overlay/ibgp/peer-groups.j2:34]: Please check the lines 42, 43. Shouldn't we have it outside the password?

Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/overlay/ibgp/peer-groups.j2:34]: Good catch! Fixed now.

**Comment by tgodaA** [comment]
Because you have fixed the logic the new artifacts are generated. Please run molecule for eos_designs_unit_tests

**Review by ClausHolbechArista — APPROVED** [review]

**Review by tgodaA — APPROVED** [review]
LGTM

**Comment by tgodaA** [comment]
I am not sure about Networking but codewise, LGTM. And its approved.
